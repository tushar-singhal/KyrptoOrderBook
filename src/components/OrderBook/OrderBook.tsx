import React from 'react';
import {connect} from 'react-redux';
import {
  AppState,
  View,
  StyleSheet,
  Text,
  FlatList,
  ListRenderItem,
  AppStateStatus,
} from 'react-native';

import {subscribeOrderBook} from '../../actions/orderBookActions';
import * as orderBookSelectors from '../../reducers/orderBookReducer';
import {Data} from 'src/types';
import * as service from '../../services/socket';
import {Header} from '../headerComponent';
import {strings} from 'src/constants/strings';
import {numberWithCommas} from '../../utils/calulations';

interface Props {
  success: object;
  error: object;
  data: Data;
  subscribeOrderBook: Function;
}

interface State {
  appState: AppStateStatus;
}

export class OrderBook extends React.PureComponent<Props, State> {
  public state: State = {
    appState: AppState.currentState,
  };

  componentDidMount() {
    const {
      props: {subscribeOrderBook},
    } = this;

    subscribeOrderBook();
    AppState.addEventListener('change', this._handleAppStateChange);
  }

  componentWillUnmount() {
    service.ws.close();
    AppState.removeEventListener('change', this._handleAppStateChange);
  }

  _handleAppStateChange = (nextAppState: AppStateStatus) => {
    this.setState({appState: nextAppState});

    if (nextAppState === 'background') {
      service.ws.close();
    }

    if (nextAppState === 'active') {
      const {subscribeOrderBook} = this.props;
      subscribeOrderBook();
    }
  };
  render() {
    const {data, error} = this.props;

    if (error) {
      return (
        <View>
          <Text>
            {strings['ERROR.TITLE']} {JSON.stringify(error)}
          </Text>
        </View>
      );
    }

    return (
      <View style={styles.parentContainer}>
        <Text style={styles.listHeader}>{strings['LABEL.BIDS']}</Text>
        <FlatList
          data={data.bids}
          renderItem={this.renderItemBids}
          keyExtractor={(item) => item[0].toString()}
          style={styles.listStyle}
          ListHeaderComponent={() => <Header />}
        />

        <Text style={styles.listHeader}>{strings['LABEL.ASKS']}</Text>

        <FlatList
          data={data.asks}
          renderItem={this.renderItemAsks}
          keyExtractor={(item) => item[0].toString()}
          style={styles.listStyle}
          ListHeaderComponent={() => <Header />}
        />
      </View>
    );
  }

  private renderItemBids: ListRenderItem<number[]> = ({item}) => (
    <View style={styles.listParent}>
      <Text style={styles.listItemBids}>{numberWithCommas(item[0].toFixed(2))} </Text>
      <Text style={styles.listItemBids}>{numberWithCommas(item[1])} </Text>
      <Text style={styles.listItemBids}>{numberWithCommas(item[2])} </Text>
    </View>
  );

  private renderItemAsks: ListRenderItem<number[]> = ({item}) => (
    <View style={styles.listParent}>
      <Text style={styles.listItemAsks}>{numberWithCommas(item[0].toFixed(2))} </Text>
      <Text style={styles.listItemAsks}>{numberWithCommas(item[1])} </Text>
      <Text style={styles.listItemAsks}>{numberWithCommas(item[2])} </Text>
    </View>
  );
}

function mapStateToProps(state: any) {
  return {
    success: orderBookSelectors.getSuccess(state),
    error: orderBookSelectors.getError(state),
    data: orderBookSelectors.getData(state),
  };
}

const actionCreators = {
  subscribeOrderBook,
};

export default connect(mapStateToProps, actionCreators)(OrderBook);

const styles = StyleSheet.create({
  listStyle: {width: '100%', flex: 1, marginTop: 20},
  listHeader: {textAlign: 'center', marginTop: 30, color: 'white', fontSize: 18},
  listItemBids: {flex: 1, textAlign: 'right', color: 'red'},
  listItemAsks: {flex: 1, textAlign: 'right', color: 'green'},
  parentContainer: {flex: 1, paddingHorizontal: 40, marginVertical: 20},
  listParent: {flex: 1, flexDirection: 'row'},
});
