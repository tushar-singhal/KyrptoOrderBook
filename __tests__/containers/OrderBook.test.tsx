import Enzyme from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import {render} from 'react-native-testing-library';

import {OrderBook} from '../../src/components/OrderBook/OrderBook';
Enzyme.configure({adapter: new Adapter()});

const createTestProps = (props?: object) => ({
  ...props,
});

jest.mock('react-redux', () => {
  return {
    connect: jest.fn().mockReturnValue(() => jest.fn()),
  };
});

jest.mock('../../src/services/socket', () => {
  return {
    ws: {
      close: jest.fn().mockReturnValue(() => jest.fn()),
    },
    close: jest.fn().mockReturnValue(() => jest.fn()),
  };
});

const dataMock = {
  bids: [
    [49272.0, 25000.0, 25000.0],
    [49275.0, 0.0, 25000.0],
  ],
  asks: [
    [49272.0, 25000.0, 25000.0],
    [49275.0, 0.0, 25000.0],
  ],
};

describe('order book', () => {
  const props: any = createTestProps({
    subscribeOrderBook: jest.fn(),
    data: dataMock,
    success: {},
    error: undefined,
  });
  const {toJSON} = render(<OrderBook {...props} />);

  it('should match snapshot', () => {
    expect(toJSON()).toMatchSnapshot();
  });
});

describe('ConnectedHome', () => {
  const mockConnect = require('react-redux').connect;
  const mapStateToProps = mockConnect.mock.calls[0][0];
  it('should map listing from state to props', () => {
    const mockState = {orderBook: {data: dataMock}};
    const props = mapStateToProps(mockState);
    expect(props.data).toEqual(dataMock);
  });
});
