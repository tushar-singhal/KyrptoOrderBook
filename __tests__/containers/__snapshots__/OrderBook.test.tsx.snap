// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`order book should match snapshot 1`] = `
<View
  style={
    Object {
      "flex": 1,
      "marginVertical": 20,
      "paddingHorizontal": 40,
    }
  }
>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 18,
        "marginTop": 30,
        "textAlign": "center",
      }
    }
  >
    BIDS
  </Text>
  <RCTScrollView
    ListHeaderComponent={[Function]}
    data={
      Array [
        Array [
          49272,
          25000,
          25000,
        ],
        Array [
          49275,
          0,
          25000,
        ],
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            PRICE
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            SIZE
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            TOTAL
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            49,272.00
             
          </Text>
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            49,275.00
             
          </Text>
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            0
             
          </Text>
          <Text
            style={
              Object {
                "color": "red",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <Text
    style={
      Object {
        "color": "white",
        "fontSize": 18,
        "marginTop": 30,
        "textAlign": "center",
      }
    }
  >
    ASKS
  </Text>
  <RCTScrollView
    ListHeaderComponent={[Function]}
    data={
      Array [
        Array [
          49272,
          25000,
          25000,
        ],
        Array [
          49275,
          0,
          25000,
        ],
      ]
    }
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
        "marginTop": 20,
        "width": "100%",
      }
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View>
      <View
        onLayout={[Function]}
      >
        <View
          style={
            Object {
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            PRICE
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            SIZE
          </Text>
          <Text
            style={
              Object {
                "color": "gray",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            TOTAL
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            49,272.00
             
          </Text>
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
        </View>
      </View>
      <View
        onLayout={[Function]}
        style={null}
      >
        <View
          style={
            Object {
              "flex": 1,
              "flexDirection": "row",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            49,275.00
             
          </Text>
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            0
             
          </Text>
          <Text
            style={
              Object {
                "color": "green",
                "flex": 1,
                "textAlign": "right",
              }
            }
          >
            25,000
             
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
